sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala

```diff
@@ -135,11 +136,19 @@ abstract class AbstractGpuJoinIterator(
     withResource(new NvtxWithMetrics(gatherNvtxName, NvtxColor.DARK_GREEN, joinTime)) { _ =>
       val ret = gathererStore.map { gather =>
         gather.checkpoint
+        gather.allowSpilling()
         withRetryNoSplit[ColumnarBatch] {
-          gather.allowSpilling()
-          gather.restore
-          val nextRows = JoinGatherer.getRowsInNextBatch(gather, targetSize)
-          gather.gatherNext(nextRows)
+          try {
+            val nextRows = JoinGatherer.getRowsInNextBatch(gather, targetSize)
+            gather.gatherNext(nextRows)
+          } catch {
+            case retryOOM: RetryOOM =>
+              // If we are retrying, we need to do this again to close the cached cb(s).
+              // This should not cause it to spill again.
+              gather.allowSpilling()
+              gather.restore
+              throw retryOOM
```

Comment:
another thing we probably need to restore is the cached buffer within `LazySpillableGatherMap`, going back to the "nothing that is not spillable logic"

Reply:
this whole thing is something Bobby pointed out earlier where we may want to incorporate in `withRetry*` a rollback function, to be invoked on OOM. Essentially the same try/catch exists there, so having a rollback policy would allow us to put the body of this into that API. Could be a follow on thing we do.. and doesn't have to be in this PR

Reply:
Agreed - I was thinking we need something like that as well.

I think the LazySpillableGatherMap should be handled by the gather.allowSpilling(), no?

Reply:
Ah I missed it. Yeah that looks to be handled.